README.md
p5-www-trello
=============

A Perl interface to the API for www.trello.com


WWW-Trello/Changes
Revision history for WWW-Trello

0.01    2014-04-19
        First version, released on an unsuspecting world.


WWW-Trello/Makefile.PL
use 5.006;
use strict;
use warnings FATAL => 'all';
use ExtUtils::MakeMaker;

WriteMakefile(
    NAME             => 'WWW::Trello',
    AUTHOR           => q{Shane Utt <shaneutt@linux.com>},
    VERSION_FROM     => 'lib/WWW/Trello.pm',
    ABSTRACT_FROM    => 'lib/WWW/Trello.pm',
    LICENSE          => 'Artistic_2_0',
    PL_FILES         => {},
    MIN_PERL_VERSION => 5.006,
    CONFIGURE_REQUIRES => {
        'ExtUtils::MakeMaker' => 0,
    },
    BUILD_REQUIRES => {
	'Want'                  => 0.23,
	'Test::Deep'            => 0.112,
        'Test::More'            => 1.001003,
	'URI'                   => 1.60,
	'LWP::UserAgent'        => 6.05,
	'HTTP::Request::Common' => 6.04,
    },
    PREREQ_PM => {
        #'ABC'              => 1.6,
        #'Foo::Bar::Module' => 5.0401,
    },
    dist  => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean => { FILES => 'WWW-Trello-*' },
);


WWW-Trello/lib/WWW/Trello/Board.pm
package WWW::Trello::Board;

use 5.006;
use strict;

use URI;
use JSON;

use WWW::Trello::Error;
use WWW::Trello::Utility;

our %ObjectMap;

=head1 NAME

WWW::Trello::Board - Board objects for WWW::Trello

=head1 SUBROUTINES/METHODS

=head2 Public Methods

=head3 new()

Create a new WWW::Trello::Board object

=cut

sub new {
	my $class = shift;
	my $params = ref($_[0]) ? $_[0] : {@_};
	my %defaults = (
		base   => 1,
		host   => 'api.trello.com',
		mode   => 'get',
		scheme => 'https',
	);

	my $self;
	map {
		$self->{$_} = $params->{$_}
	} grep {
		defined($params->{$_})
	} qw(base host mode port scheme appkey token);

	if (!$self->{appkey} or !$self->{token}) {
		WWW::Trello::Error->throw(message  => 'appkey and token must be provided.')->icantdothatdave;
	}
	my @modes = qw(delete get head post put);
	if (! grep (/$self->{mode}/i, @modes)) {
		WWW::Trello::Error->throw(
			message => 'mode must be one of the following: ' . join(', ', @modes) . '. You supplied: ' . $self->{mode},
		)->icantdothatdave;
	}
	my @schemes = qw(http https);
	if (! grep (/$self->{scheme}/i, @schemes)) {
		WWW::Trello::Error->throw(
			message  => 'scheme must be one of the following: ' . join(', ', @schemes) . '. You supplied: ' . $self->{scheme},
		)->icantdothatdave;
	}

	while (my ($key, $val) = each %defaults) {
		$self->{$key} ||= $val;
	}

	$self->{uri} = URI->new;
	$self->{uri}->scheme($self->{scheme});
	$self->{uri}->host($self->{host});
	$self->{uri}->port($self->{port}) if exists $self->{port};
	return bless $self, $class;
}

=head3 actions

Provides board actions: https://trello.com/docs/api/board/index.html#get-1-boards-board-id-actions

=cut

sub actions {
	return 'FIXME';
}

=head2 Private Methods

=head3 AUTOLOAD()

Catch unexpected method calls and be informative

=cut

sub AUTOLOAD {
	my $self = shift;
	our $AUTOLOAD;
	my ($key) = $AUTOLOAD =~ /.*::([\w_]+)/o;
	return if ($key eq 'DESTROY');
	WWW::Trello::Error->throw(message => "$key is not a valid method for " . ref($self))->icantdothatdave;
}

=head3 _call()

This method will make an eventual API call to trello's api,
and then will return expected objects.

=cut

sub _call {
	my $self = shift;
	my $args = ref($_[0]) ? $_[0] : {@_};
	my $method = delete($args->{__chain});

	if ($method && (my $call = shift(@{$method}))) {
		return $self->$call($args);
	}

	my $url;
	$url .= "/$self->{base}" if $self->{base};
	$url .= '/boards';

	my $id = delete $args->{id};
	$url .= "/$id";
	$self->{uri}->path($url);
	return $self->_httpCheck(WWW::Trello::Utility::_doRestCall($self, $self->{mode}, $self->{uri}, $args));
}

=head3 _httpCheck()

Tries the API call given via an HTTP::Request::Common,
and attempts to "do the right thing".

=cut

sub _httpCheck {
	my $self = shift;
	my $http = shift;
	if ($http->{_rc} != 200) {
		return WWW::Trello::Error->throw(
			category => 'http',
			message  => "$http->{_msg}: $http->{_content}",
			type     => $http->{rc},
		);
	} else {
		return decode_json($http->{_content}) || WWW::Trello::Error->throw(
			message  => "could not decode json response: $@",
		);
	}
}

1;

WWW-Trello/lib/WWW/Trello/Error.pm
package WWW::Trello::Error;

use 5.006;
use strict;

=head1 NAME

WWW::Trello::Error - Error handling for WWW::Trello

=head1 SUBROUTINES/METHODS

=head2 AUTOLOAD()

Die when we try to call arbitrary methods on an error object.

=cut

sub AUTOLOAD {
	my $self = shift;
	die "Error => " . $self->message;
}

=head2 throw()

Used like ->new to give an error object.

=head3 USAGE

	ARGUMENT	DESCRIPTION
	---
	type		Describes the variety of failure occurred. Potential values could be 'perl error', or
			basic http error codes such as '404', '500', e.t.c.
			~ Default: 'fail'

	category	What failed. Potential values could be 'http request', 'http auth', or something
			specific to your application like 'foobar api error'.
			~ Default: 'object' (assumes something is wrong with WWW::Trello unless told otherwise)

	message		Describes the specifics of what actually occurred.
			~ Default: 'unknown'
	---

=cut

sub throw {
	my $class = shift;
	my %args = ref($_[0]) ? %{$_[0]} : @_;
	$args{type}     = $args{type}     ? $args{type}     : 'fail';
	$args{category} = $args{category} ? $args{category} : 'object';
	$args{message}  = $args{message}  ? $args{message}  : 'unknown';
	my $self  = {
		error_type     => $args{type},
		error_category => $args{category},
		error_message  => $args{message},
	};
	return bless $self, $class;
}

=head2 category()

Helper method to retrieve the category of the current error object.

=cut

sub category {
	return shift->{error_category};
}

=head2 message()

Helper method to retrieve the message of the current error object.

=cut

sub message {
	return shift->{error_message};
}

=head2 type()

Helper method to retrieve the type of the current error object.

=cut

sub type {
	return shift->{error_type};
}

1;


WWW-Trello/lib/WWW/Trello/Utility.pm
package WWW::Trello::Utility;

use 5.006;
use strict;
use warnings FATAL => 'all';

use LWP::UserAgent;
use HTTP::Request::Common;

=head1 NAME

WWW::Trello::Utility - Utility methods for arbitrary api functionality

=cut

sub _doRestCall {
        my ($api, $method, $url, $args) = @_;
        $method = uc($method);
        $args ||= {};
        if ($api->{useragent_options} && ref($api->{useragent_options}) eq 'HASH') {
                $api->{ua} ||= LWP::UserAgent->new(%{$api->{useragent_options}});
        }
        else {
                $api->{ua} ||= LWP::UserAgent->new();
        }
        $url .= '?key=' . $api->{appkey} . '&token=' . $api->{token};
        my ($request, @params) = _generateRequest($api, $method, $url, $args);
        return $api->{ua}->request($request, @params);
}

sub _generateRequest {
        my ($api, $method, $url, $args) = @_;
        my $ua = $api->{ua};
        my @parameters = ($url, (), %{$args});
        my $parameterOffset;
        if ($method eq 'PUT'||$method eq 'POST') {
                $parameterOffset = ref($parameters[1])? 2 : 1;
        }
        else {
                $parameterOffset = 1;
        }

        my @stuff = $ua->_process_colonic_headers(\@parameters, 0);
        {
                no strict qw(refs);
                my $request = &{"HTTP::Request::Common::${method}"}(@parameters);
                return ($request, @stuff);
        }

}

1;


WWW-Trello/lib/WWW/Trello.pm
package WWW::Trello;

use 5.006;
use strict;
use warnings;

use Want;

use WWW::Trello::Board;
use WWW::Trello::Error;
use WWW::Trello::Utility;

our %SetupArgs;
our %ObjectMap;
our %SubClasses = (
	boards => 'WWW::Trello::Board',
);

=head1 NAME

WWW::Trello - A perl interface to the API for www.trello.com

=head1 VERSION

Version 0.01

=cut

our $VERSION = '0.01';

=head1 SYNOPSIS

=head1 SUBROUTINES/METHODS

=head2 new()

Create a new WWW::Trello object.

=cut

sub new {
	my $class = shift;
	my $params = ref($_[0]) ? $_[0] : {@_};
	my %defaults = (
		base   => 1,
		host   => 'api.trello.com',
		mode   => 'get',
		scheme => 'https',
	);

	map {
		$SetupArgs{$_} = $params->{$_}
	} grep {
		defined($params->{$_})
	} qw(base host mode port scheme appkey token);

	while (my ($key, $val) = each %defaults) {
		$SetupArgs{$key} ||= $val;
	}

	unless ($SetupArgs{appkey} && $SetupArgs{token}) {
		WWW::Trello::Error->throw(message  => 'appkey and token must be provided.')->icantdothatdave;
	}

	my $self = \%SetupArgs;
	bless $self, $class;
	$ObjectMap{$self} = {};

	return $self;
}

=head2 AUTOLOAD()

Builds data to make objects and api calls

=cut

sub AUTOLOAD {
	my $self = shift;

	our $AUTOLOAD;
	my ($key) = $AUTOLOAD =~ /.*::([\w_]+)/o;
	return if ($key eq 'DESTROY');

	push @{$self->{__chain}}, $key;

	my $args = ref($_[0]) ? $_[0] : {@_};
	$self->{"__$key"} = $args if (@{$self->{__chain}} == 1);
	if (want('OBJECT') || want('VOID')) {
		return $self;
	}

	my @chain = @{delete($self->{__chain})};

	my $type = shift(@chain);
	unless (grep {$_ eq $type} keys(%SubClasses)) {
		WWW::Trello::Error->throw(message => "$type is not a valid Trello API object.")->icantdothatdave;
	}

	$args->{__chain} = \@chain if @chain;
	my $id = delete($self->{"__$type"})->{id};
	unless ($id) {
		WWW::Trello::Error->throw(message  => "$type requires an id.")->icantdothatdave;
	}

	$self->_addRelatedObject($type, {id => $id});
	my $object = $self->{__lastObject} = $ObjectMap{$self}{$type}{$id};
	my $response = $object->_call($args);

	return $response;
}

=head2 last()

Returns the last object that was used

=cut

sub last {
	return shift->{__lastObject};
}

=head2 _addRelatedObject()

Stores objects under the topical object

=cut

sub _addRelatedObject {
	my $self = shift;
	my $type = shift;
	my $args = shift;
	my $buildArgs = {%SetupArgs, %{$args}};
	my $object = $SubClasses{$type}->new($buildArgs);
	$ObjectMap{$self}{$type}{$args->{id}} = $object;
	return $self;
}

=head1 AUTHOR

Shane Utt, C<< <shaneutt at linux.com> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-WWW-Trello at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=WWW-Trello>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc WWW::Trello

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=WWW-Trello>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/WWW-Trello>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/WWW-Trello>

=item * Search CPAN

L<http://search.cpan.org/dist/WWW-Trello/>

=back

=head1 LICENSE AND COPYRIGHT

Copyright (C) 2014 Shane Utt, All Rights Reserved.

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

=cut

1;
